### PR TITLE
ddl, expression: fix varchar type column flags are not compatible with MySQL

### PR DESCRIPTION
**Pull request by Rustin-Liu**
Signed-off-by: Rustin-Liu <rustin.liu@gmail.com>

### What problem does this PR solve? 

solve #13992
### What is changed and how it works?
Add flag for utf8.

### Check List 

Tests 

 - Unit test

Code changes

None
Side effects

None

Related changes

None

Release note

 None

### CODE REVIEW

**Comment by Rustin-Liu** [comment]
Hi @zimulala , It seems that my code changes have a great impact on other binary flag logics.
Am I correct on this way?   Do you have any suggestions for fix it?

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -606,7 +606,13 @@ func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef, o
 	// Set `NoDefaultValueFlag` if this field doesn't have a default value and
 	// it is `not null` and not an `AUTO_INCREMENT` field or `TIMESTAMP` field.
 	setNoDefaultValueFlag(col, hasDefaultValue)
-	if col.FieldType.EvalType().IsStringKind() && col.Charset == charset.CharsetBin {
+
+	// We need the MySQL compatibility flag, see # 13992.
+	isBinaryCollation := func(collation string) bool {
+		// @TODO need define a const for "_bin".
+		return strings.Count(collation, "_bin") == 1
```
Comment (zimulala) [inline ddl/ddl_api.go]: Try changing to "return collation == charset.CollationBin"?

Comment (Rustin-Liu) [reply ddl/ddl_api.go]: But I saw in the MySQL, if you set the collate like XXX_bin the flag always BINARY.

Comment (zimulala) [reply ddl/ddl_api.go]: I read the MySQL code and it feels like you said. I see some tests in CI fail, such as `testInferTypeSuite.TestInferType`. It needs to update this test. The test such as `testIntegrationSuite3.TestChangingCharsetToUtf8`. It seems the code we need to update in https://github.com/pingcap/tidb/blob/master/ddl/table.go#L684.
Other tests also need to be handled, if you encounter problems, you can @me.

**Pull request closed without merging**